### PR TITLE
dsp/transform: new package and initial Hilbert transform

### PR DESCRIPTION
**Pull request by robertkleffner**
First time contributor to this project. In a project I'm working on using some DSP-related code in Go, I needed to demodulate a real-valued signal to get its IQ components. The Hilbert transform is a useful way to get this information, and I was already using Gonum's FFTs elsewhere, so the implementation was straightforward and is described in numerous places on the internet.

The signal envelope is also a useful feature, so I included a method to compute that as well.

There is one TODO, regarding the necessity of normalization for the Analytic signal. So far, the results I get for both analytic signal and envelope are visually accurate, so I'm leaning towards necessary.

Justification for new `transform` folder: there are several other related signal transforms that could be eventually included here, e.g. Hilbert-Huang Transform. Originally, I had these files in the `fourier` folder, but it didn't feel right even though the implementation relies heavily on `CmplxFFT`. Open to other folder names or locations.

### CODE REVIEW

File: dsp/transform/hilbert.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+	"gonum.org/v1/gonum/dsp/fourier"
+)
+
+// Implements an approximate Hilbert transform that allows the user to determine an
+// approximate analytical signal of a real signal, and determine the real 'envelope'
+// of a signal.
+//
+// The underlying implementation uses a DFT and IDFT transform, so usual methods for
+// speeding up DFT/IDFT are likely to apply here as well.
```
Comment (kortschak) [inline dsp/transform/hilbert.go:18]:
```suggestion
// Hilbert implements an approximate Hilbert transform that allows calculation
// of an approximate analytical signal of a real signal, and determine the
// real envelope of a signal.
//
// The underlying implementation uses a discrete Fourier transform and inverse
// and discrete Fourier transform, so usual methods for speeding up these
// transforms are likely to apply here as well.
```

Is the second paragraph directed at users? If so, a link to a relevant article would be nice. If they are directed at us as maintainers, the comment should not be visible in the godoc.

File: dsp/transform/hilbert.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+// Return a new Hilbert transformer for signals of size N.
+func NewHilbert(n int) *Hilbert {
+	return &Hilbert{
+		fft:  fourier.NewCmplxFFT(n),
+		work: make([]complex128, n),
+	}
+}
+
+// Retrieve the length N of signals that are valid input for this Hilbert transform.
```
Comment (kortschak) [inline dsp/transform/hilbert.go:32]:
```suggestion
// Len returns the length of signals that are valid input for this Hilbert transform.
```

File: dsp/transform/hilbert.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+func (h *Hilbert) Len() int {
+	return len(h.work)
+}
+
+// Compute the analytical signal of a real signal, and stores the result in the dst slice,
+// returning it.
+// If the dst slice is nil, a new slice will be created and returned. The dst slice
+// must be the same length as the input signal.
+func (h *Hilbert) AnalyticSignal(signal []float64, dst []complex128) []complex128 {
```
Comment (kortschak) [inline dsp/transform/hilbert.go:41]:
```suggestion
func (h *Hilbert) AnalyticSignal(dst []complex128, signal []float64) []complex128 {
```
`dst` is conventionally the first parameter.

File: dsp/transform/hilbert.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	// forward fft of signal
+	coeff := h.fft.Coefficients(dst, h.work)
+	for i := range h.work {
+		h.work[i] = 0
+	}
+
+	// multiply positive frequencies by 2, zero out negative frequencies
+	// however, leave dc unchanged (and nyquist when n%2 == 0)
```
Comment (kortschak) [inline dsp/transform/hilbert.go:57]:
```suggestion
	// Multiply positive frequencies by 2, zero out negative frequencies.
	// However, leave dc unchanged (and nyquist when n%2 == 0).
```

File: dsp/transform/hilbert.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	h.work[0] = coeff[0]
+	for i, d := range coeff[1 : len(coeff)/2+1] {
+		h.work[i+1] = d * 2
+	}
+	if len(coeff)%2 == 0 {
+		h.work[len(coeff)/2] = coeff[len(coeff)/2]
+	}
+
+	// normalize the results for now
+	// TODO: is this important for analytic signal or envelope?
```
Comment (kortschak) [inline dsp/transform/hilbert.go:67]:
```suggestion
	// Normalize the results for now.
	// TODO: Is this important for analytic signal or envelope?
```
What is the TODO to be done?

File: dsp/transform/hilbert.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+func (h *Hilbert) AnalyticSignal(signal []float64, dst []complex128) []complex128 {
+	if dst == nil {
+		dst = make([]complex128, len(signal))
+	}
+
+	for i, v := range signal {
+		h.work[i] = complex(v, 0)
+	}
+
+	// forward fft of signal
```
Comment (kortschak) [inline dsp/transform/hilbert.go:50]:
```suggestion
	// Forward FFT of the signal.
```

File: dsp/transform/hilbert.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+// Retrieve the length N of signals that are valid input for this Hilbert transform.
+func (h *Hilbert) Len() int {
+	return len(h.work)
+}
+
+// Compute the analytical signal of a real signal, and stores the result in the dst slice,
+// returning it.
+// If the dst slice is nil, a new slice will be created and returned. The dst slice
+// must be the same length as the input signal.
```
Comment (kortschak) [inline dsp/transform/hilbert.go:40]:
```suggestion
// AnalyticSignal computes the analytical signal of a real signal, and stores
// the result in the dst slice, returning it.
//
// If the dst slice is nil, a new slice will be created and returned. The dst slice
// must be the same length as the input signal.
```

File: dsp/transform/hilbert.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	if dst == nil {
+		dst = make([]float64, len(signal))
+	}
+
+	analytic := h.AnalyticSignal(signal, nil)
+	for i, a := range analytic {
+		dst[i] = cmplx.Abs(a)
+	}
+	return dst
+}
```
Comment (kortschak) [inline dsp/transform/hilbert.go:89]: I'm not convinced this is needed. It's just a range loop.

File: dsp/transform/hilbert_test.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+func TestHilbertAnalytic(t *testing.T) {
+	testCases := []struct {
+		input []float64
+	}{
+		{[]float64{0, 0, 0, 0}},
+		{[]float64{1, 2, 3, 4}},
+		{[]float64{1, 2, 3, 4, 5}},
+	}
```
Comment (kortschak) [inline dsp/transform/hilbert_test.go:21]: Please hoist this out of the function to be a `var hilbertAnalyticSignalTests []struct{ name string; input ...`.

File: dsp/transform/hilbert_test.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("tc%d", i), func(t *testing.T) {
```
Comment (kortschak) [inline dsp/transform/hilbert_test.go:24]:
```suggestion
	for _, test := range hilbertAnalyticSignalTests {
		t.Run(test.name, func(t *testing.T) {
```
and `s/tc/test/g` elsewhere.

File: dsp/transform/hilbert_test.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+			transformer := NewHilbert(len(tc.input))
+			if transformer.Len() != len(tc.input) {
+				t.Errorf("expected Hilbert transform length to be %d, got %d", len(tc.input), transformer.Len())
+			}
```
Comment (kortschak) [inline dsp/transform/hilbert_test.go:28]:
```suggestion
			h := NewHilbert(len(tc.input))
			if h.Len() != len(test.input) {
				t.Errorf("unexpected Hilbert transform length: got:%d want:%d", h.Len(), len(test.input))
			}
```

File: dsp/transform/hilbert_test.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+			dst := make([]complex128, len(tc.input))
+			analytic := transformer.AnalyticSignal(tc.input, dst)
+			result := make([]float64, len(tc.input))
+			for i, c := range analytic {
+				result[i] = math.Round(real(c))
```
Comment (kortschak) [inline dsp/transform/hilbert_test.go:34]: This makes me queasy. Why is it being done?

File: dsp/transform/hilbert_test.go [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+			}
+			if !slices.Equal(tc.input, result) {
+				t.Errorf("expected Hilbert transform result %v, got %v", tc.input, result)
+			}
```
Comment (kortschak) [inline dsp/transform/hilbert_test.go:38]: `cmplxs.EqualApprox`?

**Review by kortschak — COMMENTED** [review]
Thanks for the contribution.

File: dsp/transform/hilbert_test.go (hunk 12 above)
Comment (robertkleffner) [reply dsp/transform/hilbert_test.go]: Originally wrote this test outside the gonum library; it is now using `cmplxs.EqualApprox`.

File: dsp/transform/hilbert.go (hunk 5 above)
Comment (robertkleffner) [reply dsp/transform/hilbert.go:68]: I think normalized is how most people will expect the analytic signal; removing TODO

File: dsp/transform/hilbert_test.go (hunk 9 above)
Comment (robertkleffner) [reply dsp/transform/hilbert_test.go]: This feedback is confusing to me, as the FFT test cases are written inside the function. With the update I made for readability, is this still needed?

File: dsp/transform/hilbert.go (hunk 8 above)
Comment (robertkleffner) [reply dsp/transform/hilbert.go:90]: My only rebuttal is that some users would only be interested in the signal envelope, and possibly have only a fuzzy idea how it relates to the analytic signal (or even that it does, as I did when I first went looking for this). They would then have to look it up in order to recreate this loop, hoping that `AnalyticSignal` leaves the results in a state that all one needs to do is `Abs` each value. (For example, Gonum's FFT does not normalize its results, which is documented but surprising). With this method, it's at least searchable that Gonum can easily be used to compute a signal envelope regardless of how much expertise one has with DSP.

As an alternative, though, perhaps I could add to the doc comment of `AnalyticSignal`: "The envelope of the signal can be straightforwardly computed from the results using `cmplx.Abs` on each element".

File: dsp/transform/hilbert.go (hunk 1 above)
Comment (robertkleffner) [reply dsp/transform/hilbert.go]: Made this comment nicer, and removed the second paragraph. I added another comment to the `NewHilbert` method noting a performance tip. My source for that performance tip is a comment [in the internal/fftpack source](https://github.com/gonum/gonum/blob/1ca563a018b641e805317f1ac9ae0d37b32d162c/dsp/fourier/internal/fftpack/cfft.go#L126)

File: dsp/transform/hilbert_test.go (hunk 9 above)
Comment (kortschak) [reply dsp/transform/hilbert_test.go]: Yes, I'd like the var outside the function.

File: dsp/transform/hilbert.go (hunk 8 above)
Comment (kortschak) [reply dsp/transform/hilbert.go]: Suggest then having this in an example test. This shows the user how to use what we have and provides the code if they want to just copy/paste.

Comment (kortschak) [reply dsp/transform/hilbert.go]: Further to this, I think a more integrated example test would be worthwhile. This would be one that works through a toy example showing how the whole thing would be used.

Also, I have just noticed that this is a new package. It should have a doc.go file with package godoc explaining what the package is intended to cover.

File: dsp/transform/hilbert_test.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+	"gonum.org/v1/gonum/cmplxs"
+)
+
+func TestHilbertAnalytic(t *testing.T) {
+	const tol = 1e-10
+	testCases := []struct {
+		in   []float64
+		want []complex128
+	}{
+		{[]float64{0, 0, 0, 0}, []complex128{0, 0, 0, 0}},
```
Comment (ccoVeille) [inline dsp/transform/hilbert_test.go:20]: Please add a test fit a nil slice, and an empty slice

This should be covered

File: dsp/transform/hilbert.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+		unnorm[i] = u / complex(float64(len(unnorm)), 0)
+	}
+	return unnorm
+}
+
+// Compute the positive envelope of a real signal, and stores the result in the dst slice,
+// returning it.
+// If the dst slice is nil, a new slice will be created and returned. The dst slice
+// must be the same length as the input signal.
+func (h *Hilbert) Envelope(dst []float64, signal []float64) []float64 {
```
Comment (ccoVeille) [inline dsp/transform/hilbert.go:77]: This method is not tested

Comment (robertkleffner) [reply dsp/transform/hilbert.go]: Removed based on additional feedback from @kortschak

File: dsp/transform/hilbert_test.go (hunk 14 above)
Comment (robertkleffner) [reply dsp/transform/hilbert_test.go]: These tests now pass, but I will note that these cases panic for the `CmplxFFT` methods that this transform is based on

File: dsp/transform/hilbert_test.go [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+var hilbertAnalyticSignalTests = []struct {
+	name string
+	in   []float64
+	want []complex128
+}{
+	{"Hilbert nil slice", nil, nil},
+	{"Hilbert empty slice", []float64{}, []complex128{}},
+	{"Hilbert zeros test", []float64{0, 0, 0, 0}, []complex128{0, 0, 0, 0}},
+	{"Hilbert whole components test", []float64{1, 2, 3, 4}, []complex128{1 + 1i, 2 - 1i, 3 - 1i, 4 + 1i}},
+	{
```
Comment (ccoVeille) [inline dsp/transform/hilbert_test.go:22]: Usually, I do a map[string]struct

Where the string is the name, it avoids adding it to the struct and make sure to provide unique name in case of copypasta

File: dsp/transform/hilbert.go [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+	fft  *fourier.CmplxFFT
+	work []complex128
+}
+
+// NewHilbert returns a new Hilbert transformer for signals of size n.
+// The transform is most efficient when n is a product of small primes.
+func NewHilbert(n int) *Hilbert {
+	if n < 1 {
+		return &Hilbert{nil, nil}
+	}
```
Comment (ccoVeille) [inline dsp/transform/hilbert.go:24]: Shouldn't it return an error? Or an empty slice still make sense

File: dsp/transform/hilbert.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+// AnalyticSignal computes the analytical signal of a real signal, and stores
+// the result in the dst slice, returning it.
+//
+// If the dst slice is nil, a new slice will be created and returned. The dst slice
+// must be the same length as the input signal, otherwise the method will panic.
+func (h *Hilbert) AnalyticSignal(dst []complex128, signal []float64) []complex128 {
+	if len(signal) != h.Len() {
+		panic("transform: input signal length mismatch")
+	}
```
[the remark this answers is not in the document]
Comment (robertkleffner) [reply dsp/transform/hilbert.go:44]: I was trying to keep this struct's API similar to the FFT API, which does not return `error`s for its compute methods, preferring to panic. The constraint is noted in the doc comment, and I can't think of a scenario where someone would reasonably expect to get an analytic signal with different size from the input, so panic felt more appropriate anyway (implementation rather than user error).

But if `error` is the desired API here, I can put it in.

File: dsp/transform/hilbert.go (hunk 17 above)
Comment (robertkleffner) [reply dsp/transform/hilbert.go:24]: Again, trying to mimic the FFT API, which `panic`s with argument `0`. I think `error` could make more sense here though.

Comment (ccoVeille) [reply dsp/transform/hilbert.go:24]: Let's wait for a maintainer then.

Nice addition, BTW

File: dsp/transform/hilbert.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply dsp/transform/hilbert.go:44]: > I would change signature to return an error and avoid panics

No. Please see https://go-talks.appspot.com/github.com/gonum/talks/2017/gonumtour.slide#14

File: dsp/transform/hilbert_test.go (hunk 16 above)
Comment (kortschak) [reply dsp/transform/hilbert_test.go:22]: > I do a map[string]struct

Please don't do this here.

File: dsp/transform/hilbert.go (hunk 17 above)
Comment (kortschak) [inline dsp/transform/hilbert.go:24]: Please remove this and allow a panic. A negative would be indicative of failure to hold a known invariant.

File: dsp/transform/hilbert.go [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+		dst = make([]complex128, len(signal))
+	} else if len(dst) != h.Len() {
+		panic("transform: destination length mismatch")
+	}
+
+	// edge case where the length is zero or less
+	if h.fft == nil {
+		return dst
+	}
+
```
Comment (kortschak) [inline dsp/transform/hilbert.go:56]: This is not needed.
```suggestion
```

File: dsp/transform/hilbert_example_test.go [hunk 20]
```diff
@@ -0,0 +1,46 @@
+package transform
+
+import (
+	"fmt"
+	"math/cmplx"
+)
+
+func ExampleHilbert_AnalyticSignal() {
+	// Samples is a set of real amplitudes that make up a signal
```
Comment (kortschak) [inline dsp/transform/hilbert_example_test.go:9]:
```suggestion
	// Samples is a set of real amplitudes that make up a signal.
```

File: dsp/transform/hilbert_example_test.go [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+	samples := []float64{1, 0, 2, 0, 4, 0, 2, 0}
+
+	// Initialize a Hilbert transform and 'demodulate' to get the analytic signal.
+	// The result is the complex I/Q (In-Phase / Quadrature) demodulation of the input signal.
```
Comment (kortschak) [inline dsp/transform/hilbert_example_test.go:13]: These comments should be shorter so that they render nicely in the godoc; the don't get wrapped in the presented code, so we need to do that work here.
```suggestion
	// Initialize a Hilbert transform and 'demodulate' to get the
	// analytic signal.
	// The result is the complex I/Q (In-Phase / Quadrature) demodulation
	// of the input signal.
```

File: dsp/transform/hilbert_example_test.go [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+	h := NewHilbert(len(samples))
+	iqSamples := h.AnalyticSignal(nil, samples)
+
+	// We can compute the instantaneous amplitude of the signal (or 'envelope') using absolute value.
+	// Analyzing the envelope is an easy way to measure changes in amplitude over time in a signal.
```
Comment (kortschak) [inline dsp/transform/hilbert_example_test.go:18]:
```suggestion
	// We can compute the instantaneous amplitude of the signal
	// (or 'envelope') using absolute value. Analyzing the envelope
	// is an easy way to measure changes in amplitude over time in a
	// signal.
```

File: dsp/transform/hilbert_example_test.go [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+	envelope := make([]float64, len(samples))
+	for ind, iq := range iqSamples {
+		envelope[ind] = cmplx.Abs(iq)
+	}
+
+	// We can also compute the instantaneous phase of each part of the signal using the 4-quadrant arc-tangent.
+	// With multiple samples, the instantaneous phase can be used to estimate instantaneous frequency of a signal.
```
Comment (kortschak) [inline dsp/transform/hilbert_example_test.go:25]:
```suggestion
	// We can also compute the instantaneous phase of each part of the
	// signal using the 4-quadrant arc-tangent. With multiple samples,
	// the instantaneous phase can be used to estimate instantaneous
	// frequency of a signal.
```

File: dsp/transform/hilbert_test.go [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+
+	"gonum.org/v1/gonum/cmplxs"
+)
+
+var hilbertAnalyticSignalTests = []struct {
+	name string
+	in   []float64
+	want []complex128
+}{
+	{"Hilbert nil slice", nil, nil},
```
Comment (kortschak) [inline dsp/transform/hilbert_test.go:18]:
```suggestion
	{
		name: "nil_slice",
		in:   nil,
		want: nil,
	},
```
We know that this is a Hilbert test, so we don't need that. The field names make things more navigable. Using spaces in the name means that when the test output is emitted, the name can be directly searched for in the source (`go test` does `s/ /_/g` in test names). Similar changes below.

File: dsp/transform/hilbert_test.go [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+		t.Run(test.name, func(t *testing.T) {
+			h := NewHilbert(len(test.in))
+			if h.Len() != len(test.in) {
+				t.Errorf("unexpected Hilbert transform length: got:%d, want:%d", h.Len(), len(test.in))
+			}
+
+			dst := make([]complex128, len(test.in))
+			analytic := h.AnalyticSignal(dst, test.in)
+			if !cmplxs.EqualApprox(test.want, analytic, tol) {
+				t.Errorf("expected Hilbert transform result %v, got %v", test.want, analytic)
```
Comment (kortschak) [inline dsp/transform/hilbert_test.go:48]:
```suggestion
			got := h.AnalyticSignal(dst, test.in)
			if !cmplxs.EqualApprox(got, test.want, tol) {
				t.Errorf("unexpected Hilbert transform result:\ngot: %v\nwant:%v", got, test.want)
```

File: dsp/transform/doc.go [hunk 26]
```diff
@@ -0,0 +1,6 @@
+// Copyright ©2024 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package transform provides important transforms on signals used in digital signal processing
```
Comment (kortschak) [inline dsp/transform/doc.go:5]:
```suggestion
// Package transform provides important transforms on signals used in digital signal processing.
```

**Review by kortschak — COMMENTED** [review]
Should be good after this. Thanks.

File: dsp/transform/hilbert_test.go (hunk 16 above)
Comment (ccoVeille) [reply dsp/transform/hilbert_test.go:22]: OK.

I'm fine with having no map[string]struct but do you have any reason you could share

File: dsp/transform/hilbert.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (ccoVeille) [reply dsp/transform/hilbert.go:44]: OK.

File: dsp/transform/hilbert_test.go (hunk 16 above)
Comment (kortschak) [reply dsp/transform/hilbert_test.go:22]: Consistent test ordering.

Comment (ccoVeille) [reply dsp/transform/hilbert_test.go:22]: I have never had a need for that, except an `//Output:` in an `ExampleFoo`

But OK, I'm fine with this

File: dsp/transform/hilbert_example_test.go [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+	phase := make([]float64, len(samples))
+	for ind, iq := range iqSamples {
+		phase[ind] = cmplx.Phase(iq)
+	}
+
+	for i, iq := range iqSamples {
+		fmt.Printf("ind=%d -> I=%.4g, Q=%.4g, envelope=%.4g, phase=%.4g\n",
```
Comment (kortschak) [inline dsp/transform/hilbert_example_test.go:32]:
```suggestion
		fmt.Printf("ind=%d -> I=%.4f, Q=%.4f, envelope=%.4f, phase=%.4f\n",
```
`%g` offers the opportunity to represent ~0 as the tiny value, which we probably don't want.

File: dsp/transform/hilbert.go [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+// of an approximate analytical signal of a real signal, and determine the
+// real envelope of a signal.
+type Hilbert struct {
+	fft  *fourier.CmplxFFT
+	work []complex128
+}
+
+// NewHilbert returns a new Hilbert transformer for signals of size n.
+// The transform is most efficient when n is a product of small primes.
+// If n is 0 or negative, the function will panic.
```
Comment (kortschak) [inline dsp/transform/hilbert.go:21]:
```suggestion
// n must not be less than one.
```
I would prefer to state what a user cannot do than what we guarantee to do.

(though I don't think we make this statement in the cause for the panic; I need to think about how I actually want to handle that)

File: dsp/transform/hilbert_example_test.go [hunk 29]
```diff
@@ -0,0 +1,52 @@
+package transform
```
Comment (kortschak) [inline dsp/transform/hilbert_example_test.go:1]: We need a copyright notice here.

**Review by kortschak — APPROVED** [review]
This looks good. Thanks.

Please now send **another PR** adding yourself to AUTHORS and CONTRIBUTORS with the commit message, "A+C: add Robert Kleffner". When that's done and approved, I'll merge both.

**Comment by kortschak** [comment]
No need to update the branch; we'll squash what's here. Don't undo this though.
